System/Random/Internal.hs

```diff
@@ -688,20 +681,66 @@ foreign import prim "stg_word64ToDoubleyg"
     stgWord64ToDouble :: Word64# -> Double#
 #endif
 
-
 instance UniformRange Float where
   uniformRM (l, h) g = do
-    w32 <- uniformWord32 g
-    let x = word32ToFloatInUnitInterval w32
+    x <- floatInUnitIntervalM g
     return $ (h - l) * x + l
 
--- | Turns a given uniformly distributed 'Word32' value into a uniformly
--- distributed 'Float' value in the range [0,1).
-word32ToFloatInUnitInterval :: Word32 -> Float
-word32ToFloatInUnitInterval w32 = between1and2 - 1.0
+-- | Counts the number of failed Bernoulli trials with p=0.5 before the first
+-- success, up to the given limit.
+geometricDistr :: MonadRandom g s m => Int -> Int -> g s -> m Int
+geometricDistr start limit g = go start
   where
-    between1and2 = castWord32ToFloat $ (w32 `unsafeShiftR` 9) .|. 0x3f800000
-{-# INLINE word32ToFloatInUnitInterval #-}
+    go !acc
+      | acc >= limit = return limit
+      | otherwise = do
+          w <- uniformWord64 g
+          if w /= 0
+            then return $ acc + countLeadingZeros w
+            else go (acc + finiteBitSize w)
+{-# INLINE geometricDistr #-}
+
+-- | Generates a 'Float' in [0,1].
+floatInUnitIntervalM :: MonadRandom g s m => g s -> m Float
+floatInUnitIntervalM g = do
+  let maxExp = 126 -- exponent of 1.0, decremented by one
+  let mantissaBits = 23
+  let bernoulliBits = 64 - mantissaBits - 1
+  let carryMask = 2 ^ mantissaBits
+  let mantissaMask = (2 ^ mantissaBits) - 1
+
+  w <- uniformWord64 g
+  let m = w .&. mantissaMask
+  let carry = if (m /= 0) then 0 else ((w .&. carryMask) `unsafeShiftR` mantissaBits)
+
+  d <- case countLeadingZeros w of
+    b | b < bernoulliBits -> return b
+    _ -> geometricDistr bernoulliBits maxExp g
+  let e = fromIntegral (maxExp - d) + carry
+
+  return $ castWord32ToFloat $ fromIntegral $ (e `unsafeShiftL` mantissaBits) .|. m
+{-# INLINE floatInUnitIntervalM #-}
+
+-- | Generates a 'Double' in [0,1].
+doubleInUnitIntervalM :: MonadRandom g s m => g s -> m Double
+doubleInUnitIntervalM g = do
```

Comment:
This is almost identical to `floatInUnitIntervalM`. I'd recommend reducing this duplication by creating a common function where things that are different are supplied as arguments.

Reply:
Good point. This should be generalisable.

Reply:
Done.